### PR TITLE
2.x: Add singleOrError, firstOrError, lastOrError & elementAtOrError to Observable and Flowable

### LINKED ISSUE #4588: 2.x: Convert Observable to Single without a default value
With the RC3 changes I don't see an easy way of converting an Observable into a Single without giving a default value.

How do you feel about adding methods for that?

### PR DESCRIPTION
**Pull request by vanniktech**
Closes #4588

### CODE REVIEW

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableElementAtSingle.java [hunk 1]
```diff
@@ -104,7 +105,12 @@ public void onComplete() {
             s = SubscriptionHelper.CANCELLED;
             if (index <= count && !done) {
                 done = true;
-                actual.onSuccess(defaultValue);
+
+                if (defaultValue != null) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableElementAtSingle.java]:
``` java
T v = defaultValue;
if (v != null) {
   actual.onSuccess(v);
} else {
```

File: src/main/java/io/reactivex/internal/operators/observable/ObservableElementAtSingle.java [hunk 2]
```diff
@@ -96,7 +99,11 @@ public void onError(Throwable t) {
         public void onComplete() {
             if (index <= count && !done) {
                 done = true;
-                actual.onSuccess(defaultValue);
+                if (defaultValue != null) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/observable/ObservableElementAtSingle.java]: Read into `T v`

**Review by akarnokd — APPROVED** [review]
Great job!

**Comment by akarnokd** [comment]
There are javadoc mistakes btw.

**Comment by vanniktech** [comment]
Yup I saw them and I'll fix them up. I didn't execute that test locally. (running all tests take forever on my local machine)

**Comment by akarnokd** [comment]
Oh, this broke because of the error fixes.

/cc @JakeWharton and @abersnaze on naming/functionality.

**Review by PraveenNanda124 — COMMENTED** [review]
Great job

**Pull request merged**